incipyt/_internal/utils.py

```diff
@@ -0,0 +1,143 @@
+import collections.abc
+from string import Formatter
+
+import click
+
+
+class Requires:
+    def __init__(self, template, confirmed=False, sanitizer=None, **kwargs):
+        self._confirmed = confirmed
+        self._sanitizer = sanitizer
+        self._template = template
+        self._kwargs = kwargs
+
+    def __str__(self):
+        return f"f'{self._template}'"
+
+    def __call__(self, environment):
+        for _, key, _, _ in Formatter().parse(self._template):
+            if key not in self._kwargs:
+                continue
+
+            environment.push(key, self._kwargs[key], confirmed=self._confirmed)
+
+        args = {
+            key: (
+                self._sanitizer(key, environment.pull(key))
+                if self._sanitizer
+                else environment.pull(key)
+            )
+            for _, key, _, _ in Formatter().parse(self._template)
+            if key is not None
+        }
+        if any(not v for v in args.values()):
+            return None
+
+        return self._template.format(**args)
+
+
+class MultipleValues:
+    def __init__(self, head, tail):
+        self._values = (
+            [head] + tail._values if isinstance(tail, MultipleValues) else [head, tail]
+        )
+
+    def __call__(self, environment):
+        return click.prompt(
+            "Conflicting configuration, choose between",
+            type=click.Choice(
+                [
+                    value(environment) if callable(value) else value
+                    for value in self._values
+                ]
+            ),
+        )
+
+    def __repr__(self):
+        return f"{type(self).__name__}({self._values})"
+
+    def __eq__(self, other):
+        try:
+            return self._values == other._values
+        except AttributeError:
+            return False
+
+
+def append_unique(config, value):
+    assert isinstance(config, collections.abc.MutableSequence)
+
+    if value not in config:
+        config.append(value)
+
+
+def is_not_string_sequence(obj):
+    return (
+        isinstance(obj, collections.abc.Sequence)
+        and not isinstance(obj, collections.abc.ByteString)
+        and not isinstance(obj, str)
+    )
+
+
+class TemplateDict(collections.UserDict):
```

Comment:
```suggestion
class TemplateDict(collections.UserDict):
    def __init__(self, data):
        self.data = data

    def __setitem__(self, keys, values):
        if isinstance(keys, str) or keys is None:
            self[(keys,)] = values
            return

        if not isinstance(values, tuple):
            self[keys] = values, Requires
            return

        value, function = values

        if isinstance(value, collections.abc.Mapping):
            for k, v in value.items():
                self[keys + (k,)] = v, function
            return

        config = self.data
        for key in keys[:-1]:
            if key not in config:
                config[key] = {}
            config = config[key]

        if isinstance(value, str) or callable(value):
            value = value if callable(value) else function(value)
            if keys[-1] in config:
                if config[keys[-1]] != value:
                    config[keys[-1]] = MultipleValues(value, config[keys[-1]])
            else:
                config[keys[-1]] = value
        elif isinstance(value, collections.abc.Sequence):
            if keys[-1] not in config:
                config[keys[-1]] = []
            assert isinstance(
                config[keys[-1]], collections.abc.Sequence
            ), f"{keys} is already a mappable, cannot set to a sequence."
            for v in [
                v_ if callable(v_) else function(v_)
                for v_ in value
                if v_ not in config[keys[-1]]
            ]:
                config[keys[-1]].append(v)

    def __ior__(self, other):
        if not isinstance(other, tuple):
            self |= other, Requires
            return

        mapping, function = other

        assert isinstance(
            mapping, collections.abc.Mapping
        ), f"RHS of |= for {type(self)} should be a mapping."
        for key, value in mapping.items():
            self[key] = value, function

        return self

    def __or__(self, other):
        raise NotImplementedError(
            f"{type(self)} do not support |, use self.data | or |=."
        )

```
- Support to include `append_unique`
- Support for `conf["key1", "key2"] = #value, dict or array`
- Support for `conf |= dict`
- Support for `conf[...] = value, transformer_function`